test/functional/validateStateTransition.spec.js

```diff
@@ -0,0 +1,277 @@
+
+    dataContract.contractId = userId;
+
+    documents.forEach((d) => {
+      // eslint-disable-next-line
+      d.contractId = userId;
+      // eslint-disable-next-line
+      d.userId = userId;
+    });
+  });
+
+  it('should validate contract state transition without a blockchain user', async () => {
+    dataContract.contractId = Buffer.alloc(32).toString('hex');
+
+    const stateTransition = new DataContractStateTransition(dataContract);
+
+    await withinTransaction(async (blockHeight, blockHash) => {
+      const request = new ApplyStateTransitionRequest();
+      request.setStateTransition(stateTransition.serialize());
+      request.setBlockHeight(blockHeight);
+      request.setBlockHash(blockHash);
+
+      try {
+        await driveUpdateStateApi.applyStateTransition(request);
+        expect.fail('Error was not thrown');
+      } catch (e) {
+        const [error] = JSON.parse(e.metadata.get('errors')[0]);
+        expect(error.message).to.equal('User not found');
+        expect(e.message).to.equal('3 INVALID_ARGUMENT: Invalid argument: Invalid State Transition');
+      }
+    });
+  });
+
+  it('should validate contract state transition when it submitted twice', async () => {
+    const stateTransition = new DataContractStateTransition(dataContract);
+
+    await withinTransaction(async (blockHeight, blockHash) => {
+      const request = new ApplyStateTransitionRequest();
+      request.setStateTransition(stateTransition.serialize());
+      request.setBlockHeight(blockHeight);
+      request.setBlockHash(blockHash);
+
+      await driveUpdateStateApi.applyStateTransition(request);
+    });
+
+    await withinTransaction(async (blockHeight, blockHash) => {
+      const request = new ApplyStateTransitionRequest();
+      request.setStateTransition(stateTransition.serialize());
+      request.setBlockHeight(blockHeight);
+      request.setBlockHash(blockHash);
+
+      try {
+        await driveUpdateStateApi.applyStateTransition(request);
+      } catch (e) {
+        const [error] = JSON.parse(e.metadata.get('errors')[0]);
+        expect(error.message).to.equal('Data Contract is already present');
+      }
+    });
+  });
+
+  it('should validate document uniqueness by using indicies', async () => {
+    const [,,, indexDocument, anotherDocument] = documents;
+
+    anotherDocument.set('lastName', 'Birkin');
+
+    const stateTransition = new DataContractStateTransition(dataContract);
+
+    await withinTransaction(async (blockHeight, blockHash) => {
+      const request = new ApplyStateTransitionRequest();
+      request.setStateTransition(stateTransition.serialize());
+      request.setBlockHeight(blockHeight);
+      request.setBlockHash(blockHash);
+
+      await driveUpdateStateApi.applyStateTransition(request);
+    });
+
+    const documentsStateTransition = new DocumentsStateTransition([indexDocument]);
+
+    await withinTransaction(async (blockHeight, blockHash) => {
```

Comment:
Let's check in the same transaction? In this case, we will check transaction isolation as well.